### PR TITLE
Resolve #140 via USDA crop normalization and complete api_name transition groundwork (#156, #157 - work done by Mei)

### LINKED ISSUE #140: [Bug]: API endpoints require exact casing and whitespace for crop/resource name lookups
### Pre-flight checklist

- [x] I have searched the existing issues
- [x] I have read the contributing guidelines

### Describe the bug

The webservice API endpoints perform exact string matching on crop and resource name path parameters. This means users must know the precise casing and internal whitespace of names stored in the database to get a successful response.

For example, the USDA commodity `CORN  ALL` (two spaces, uppercase) is only matched by the exact string `CORN  ALL`. Passing `corn all`, `Corn All`, or `CORN ALL` (single space) all return a 404 `CropNotFoundException`.

Similarly, resource names like `almond hulls` require exact lowercase with spaces — `Almond Hulls`, `ALMOND HULLS`, or `almond_hulls` all fail.

This is a poor API consumer experience since:
- Users have no way to know that `CORN  ALL` has two spaces without inspecting the database
- The double-space comes from the upstream USDA NASS API and is an artifact, not meaningful data
- Casing should not matter for REST path parameters representing names

### Steps to reproduce

1. Start the webservice: `pixi run start-webservice`
2. Open Swagger UI at `http://127.0.0.1:8000/docs`
3. Try the Census endpoint `GET /v1/feedstocks/usda/census/crops/{crop}/geoid/{geoid}/parameters`
4. Enter `corn all` (lowercase, single space) for crop, `06047` for geoid
5. Response: `404 — Crop not found in database: corn all`
6. Enter `CORN  ALL` (uppercase, **two** spaces) for crop
7. Response: `200 OK`

Same issue applies to all resource-based endpoints (availability, analysis, census-by-resource, survey-by-resource).

### Expected behavior

The API should perform case-insensitive and whitespace-normalized lookups for crop and resource name parameters. All of the following should resolve to the same commodity:

- `corn all`
- `CORN ALL`
- `Corn  All`
- `CORN  ALL`

### Relevant log output / screenshots

```shell
## Affected services

- `usda_census_service.py` — `_get_commodity_by_name()` and `_get_commodity_by_resource()`
- `usda_survey_service.py` — `_get_commodity_by_name()` and `_get_commodity_by_resource()`
- `availability_service.py` — `_get_resource_by_name()`
- `analysis_service.py` — `_get_resource_by_name()`

## Proposed fix

Normalize user input (collapse whitespace, lowercase) and use `func.lower(func.regexp_replace(...))` on the DB column for comparison, so lookups are case- and whitespace-insensitive.
```

### Software version

v0.1.0 (current main)

### Environment

None

### Other environment details

_No response_

### PR DESCRIPTION
**Pull request by Vraj1234**
## 📄 Description

Issues Resolved in this : #140 , #156 , #157 

This PR is a scoped extraction from Mei's PR #148 plus a focused part 2 update for USDA crop lookup behavior.

Part 1 scope retained from #148:
1. Lowercase normalization in commodity mapping seeding for `commodity_name`, `api_name`, and `resource_name`. #156 
2. USDA census/survey view output now uses `api_name` (lowercased) instead of `name` for `usda_crop`. #157 
3. Forward migration added to refresh USDA views safely without rewriting migration history.

Part 2 updates:
1. Contain cascading effects of `name -> api_name` by making USDA crop resolution explicitly prefer `api_name` with fallback to legacy `name`.
2. Resolve issue #140 matching behavior for crop input as exact, case-insensitive, and space-insensitive.
3. Scope normalization only to crop lookups in USDA census/survey services, not resource/parameter routes.

Matching behavior decision:
1. We intentionally discarded wildcard/prefix matching.
2. These endpoints represent a single crop selector and should not return data for multiple crops.
3. Example: `corn all` should match `CORN ALL`, `Corn   All`, etc., but must not match `corn` or `corn altogether`.

App-side vs SQL-side processing discussion and decision:
1. Previous app-side approach normalized in Python and effectively scanned candidate commodities in memory (`O(N)` with row transfer to app).
2. New approach performs normalized exact comparison in SQL and keeps logic at the DB query layer.
3. Current SQL still has `O(N)` worst-case behavior because normalization is applied as an expression on columns without a dedicated index.
4. We still chose SQL-side now because it reduces application-side scanning and keeps comparison semantics centralized/consistent.
5. Long-term target for near `O(log N)` is an indexed normalized key (persisted normalized column or expression index).

Why no `ILIKE`:
1. We normalize both sides to canonical lowercase + collapsed spaces and then use exact equality.
2. For exact matching after normalization, `=` is sufficient and clearer than `ILIKE`.
3. Avoiding wildcard semantics was a deliberate requirement for this endpoint behavior.

Explicitly excluded from #148 in this PR:
1. `observation.record_id` cast/join rewrites.
2. `pixi.toml` `refresh-views` `DATABASE_URL` changes.
3. Alembic `down_revision` rewiring in historical migrations.
4. Unrelated docs/symlink/newline noise.

## ✅ Checklist

- [x] I ran `pre-commit run --all-files` and all checks pass
- [x] Tests added/updated where needed
- [ ] Docs added/updated if applicable
- [x] I have linked the issue this PR closes (if any)

## 🔗 Related Issues

Resolves #140

## 💡 Type of change

| Type             | Checked? |
| ---------------- | -------- |
| 🐞 Bug fix       | [x]      |
| ✨ New feature   | [x]      |
| 📝 Documentation | [ ]      |
| ♻️ Refactor      | [x]      |
| 🛠️ Build/CI      | [ ]      |
| Other (explain)  | [ ]      |

## 🧪 How to test

1. `pixi install`
2. `pixi run start-services`
3. `pixi run service-status`
4. `pixi run migrate`
5. `pixi run refresh-views`
6. `pixi run access-db -c "SELECT COUNT(*) AS non_lower_census FROM ca_biositing.usda_census_view WHERE usda_crop <> LOWER(usda_crop); SELECT COUNT(*) AS non_lower_survey FROM ca_biositing.usda_survey_view WHERE usda_crop <> LOWER(usda_crop);"`
7. `pixi run python -m pytest --verbose tests/webservice/v1/test_usda_census.py tests/webservice/v1/test_usda_survey.py`
8. `pixi run test`
9. `pixi run pre-commit-all`

Expected:
- Migration applies cleanly through `b6aa2fc6cd42`
- USDA views refresh successfully
- `non_lower_census = 0` and `non_lower_survey = 0`
- USDA crop endpoints match exact crop names case-insensitively and space-insensitively
- `corn all` variants match, but `corn` does not match `corn all`
- Full test suite passes

## 📝 Notes to reviewers

- This remains a scoped extraction, not a full port of #148.
- Part 2 intentionally updates only USDA crop matching paths in:
  - `usda_census_service.py`
  - `usda_survey_service.py`
- Resource and parameter route matching semantics are intentionally unchanged.

## Release Notes

* **New Features**
  * USDA Census and Survey crop name lookups now handle case variations and extra whitespace automatically for more flexible data retrieval
  * API-based commodity identifiers are now prioritized over legacy names for improved accuracy

* **Bug Fixes**
  * Enhanced commodity name normalization to prevent lookup failures caused by minor formatting differences

### CODE REVIEW

**Comment by Vraj1234** [comment]
@mglbleta part 1 of this PR is your work, please let me know if you think there are any changes:

Part 1 scope retained from https://github.com/uw-ssec/ca-biositing/pull/148:
- Lowercase normalization in commodity mapping seeding for commodity_name, api_name, and resource_name.
- USDA census/survey view output now uses api_name (lowercased) instead of name for usda_crop.
- Forward migration added to refresh USDA views safely without rewriting migration history.

File: alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    op.execute(f"DROP MATERIALIZED VIEW IF EXISTS {VIEW_SCHEMA}.usda_census_view")
+    op.execute(f"DROP MATERIALIZED VIEW IF EXISTS {VIEW_SCHEMA}.usda_survey_view")
+
+    _create_view("usda_census_view", USDA_CENSUS_VIEW)
+    _create_view("usda_survey_view", USDA_SURVEY_VIEW)
+
+
+def downgrade() -> None:
+    """Downgrade schema."""
+    op.execute(f"DROP MATERIALIZED VIEW IF EXISTS {VIEW_SCHEMA}.usda_census_view")
```
Comment (lsetiawan) [inline alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py]:
```suggestion
    op.execute(f"DROP MATERIALIZED VIEW IF EXISTS {VIEW_SCHEMA}.usda_census_view CASCADE")
```

**Comment by lsetiawan** [comment]
## PR #155 Review: USDA Crop Normalization & api_name Transition

### Overview
This PR resolves #140 by adding case-insensitive, space-insensitive exact crop matching to USDA census/survey endpoints. It also transitions views from `uc.name` to `LOWER(uc.api_name)` and adds a forward migration to refresh those materialized views.

The overall approach is well-reasoned and the PR description is excellent in documenting design decisions. However, there are several issues to address before merging.

---

### HIGH Priority

**1. Full code duplication across census and survey services**

`_normalize_crop_name`, `_normalized_sql_text`, and the entire `_get_commodity_by_name` method are copy-pasted identically in both `usda_census_service.py` and `usda_survey_service.py`. Any future bug fix or normalization change must be applied in two places.

**Recommendation:** Extract to a shared module (e.g., `services/usda_common.py`):

```python
# services/usda_common.py
def normalize_crop_name(crop_name: Optional[str]) -> str:
    if not crop_name:
        return ""
    return " ".join(crop_name.split()).lower()

def normalized_sql_text(column):
    normalized = column
    for _ in range(8):
        normalized = func.replace(normalized, "  ", " ")
    return func.lower(func.trim(normalized))

def get_commodity_by_name(session: Session, crop_name: str) -> UsdaCommodity:
    # ... shared implementation
```

---

### MEDIUM Priority

**2. `_normalized_sql_text` uses 8 iterations of `REPLACE` -- performance and correctness concerns**

Each query generates 8 nested `func.replace()` calls applied to every row (full table scan). While the `usda_commodity` table is likely small (<1000 rows), this is wasteful and the approach has an asymmetry with the Python side:

- **Python**: `" ".join(s.split())` handles *unlimited* consecutive spaces in O(n)
- **SQL**: 8 iterations of `REPLACE("  ", " ")` handles at most 2^8 = 256 consecutive spaces

**Recommendation:** Use PostgreSQL's `regexp_replace` for single-pass normalization:
```python
func.lower(func.trim(func.regexp_replace(column, r'\s+', ' ', 'g')))
```
If SQLite test compatibility is needed, dispatch at runtime or use `regexp_replace` only in production with a simpler fallback for tests. Alternatively, since the seeding pipeline already normalizes data on ingest (`str.strip().str.lower()`), the DB data should already be clean -- meaning the 8-iteration SQL normalization is purely defensive for the *input* side. Document this assumption.

**3. Two sequential queries double worst-case latency**

`_get_commodity_by_name` issues query 1 against `api_name`, then on miss, query 2 against `name`. This can be combined:

```python
stmt = (
    select(UsdaCommodity)
    .where(or_(
        and_(UsdaCommodity.api_name.is_not(None), api_expr == normalized_query),
        and_(UsdaCommodity.name.is_not(None), name_expr == normalized_query),
    ))
    .order_by(
        case((and_(UsdaCommodity.api_name.is_not(None), api_expr == normalized_query), 0), else_=1),
        UsdaCommodity.id,
    )
    .limit(1)
)
```

This preserves api_name preference while halving DB round trips.

**4. Empty crop name executes two wasted queries before raising 404**

`_normalize_crop_name(None)` returns `""`, which proceeds to run both DB queries before eventually raising `CropNotFoundException`. Add an early guard:

```python
normalized_query = normalize_crop_name(crop_name)
if not normalized_query:
    raise CropNotFoundException(crop_name or "")
```

---

### LOW Priority

**5. Missing type annotation on `column` parameter**

`_normalized_sql_text(column)` lacks a type hint. Should be:
```python
from sqlalchemy.sql.elements import ColumnElement

def _normalized_sql_text(column: ColumnElement) -> ColumnElement:
```

**6. No expression index (acknowledged as future work)**

The PR description correctly notes the current `O(N)` scan and targets an indexed normalized key long-term. For reference, the concrete migration would be:
```sql
CREATE INDEX idx_usda_commodity_api_name_norm
ON usda_commodity (LOWER(TRIM(regexp_replace(api_name, '\s+', ' ', 'g'))));
```

**7. Test coverage gaps**

The existing tests cover the key scenarios well (case-insensitive, space-collapsing, no prefix match, fallback to `name` column). Missing edge cases:
- Leading/trailing whitespace: `"  corn  "` should match `"corn"`
- Whitespace-only input: `"   "` should 404
- Unit tests for `_normalize_crop_name` in isolation

---

### What looks good

- Migration approach is sound: DROP/CREATE for materialized views with correct downgrade SQL frozen in time
- Test fixture design properly exercises both the `api_name` path (corn) and `name` fallback path (soybeans with `api_name=None`)
- The `seed_commodity_mappings.py` normalization ensures clean data at ingest time
- Intentional exclusion of ILIKE/wildcard semantics is well-documented and correct for the use case
- View changes (`LOWER(uc.api_name)`) in both SQL and Python view definitions are consistent

---

### Summary

| Priority | Issue | Action |
|----------|-------|--------|
| **HIGH** | Code duplication (3 methods x 2 files) | Extract to `usda_common.py` |
| **MEDIUM** | 8x `REPLACE` loop: perf + 256-space limit | Use `regexp_replace` or document limit |
| **MEDIUM** | Two sequential queries | Combine into single `OR` + `CASE ORDER BY` |
| **MEDIUM** | Empty input runs wasted queries | Add early guard |
| **LOW** | Missing type hint on `column` | Add `ColumnElement` annotation |
| **LOW** | No expression index | Document as follow-up ticket |
| **LOW** | Edge case test gaps | Add whitespace-edge tests |

The PR is solid in its core design. The main blockers are the code duplication (HIGH) and the normalization robustness issues (MEDIUM). The rest are improvements that could be addressed in follow-up.

File: alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py (hunk 1 above)
Comment (Vraj1234) [reply alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py]: Accepted change

File: alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py
[the remark this answers is not in the document]
Comment (Vraj1234) [reply alembic/versions/b6aa2fc6cd42_refresh_usda_views_to_use_api_name.py]: `refresh_all_views()` only refreshes existing materialized view data; it does not change view definitions. This migration changes the view definition (`name` -> `lower(api_name)`), so drop/create is required for correctness.

the main purpose is to change the name

File: resources/sql/create_analytical_views.sql
[the remark this answers is not in the document]
Comment (Vraj1234) [reply resources/sql/create_analytical_views.sql]: this file is not used by the runtime migration/refresh path. After confirming with Mei

I removed `resources/sql/create_analytical_views.sql` to avoid duplicate/ambiguous sources of truth; authoritative definitions are in `datamodels/views.py` and Alembic migrations.

**Comment by Vraj1234** [comment]
As for the other comments:

I addressed the priority items (1–3) in this PR:

1) Normalization performance: replaced 8-pass whitespace collapse with a shared 3-pass normalization helper (keeps current SQLite test compatibility).
2) Fallback query shape: merged api_name + name matching into a single query with `CASE` ordering so api_name matches are preferred.
3) Duplication: extracted shared crop normalization + commodity lookup logic into a common USDA helper module used by both census and survey services.

I also applied migration/thread feedback (`CASCADE` drops + migration rationale comments) and fixed the `get_local_engine()` bug in `seed_commodity_mappings.py`.

For items 4–5 (normalized expression indexes + normalized uniqueness enforcement), I kept this PR scoped and did not touch them. they are enhancements. Do let me know if i should create issues for them. They seem to issues we can solve if we have extra time. Its a performance optimization. 

Also, when I work on the next PR about fetching API data from views, I will try to make sure I can do something like this.

**Comment by Vraj1234** [comment]
Hi @mglbleta, this error is most likely because your code is not able to find the DATABASE_URL

The easiest solution is, in the root of your project please make a `.env` file and paste this in. 

```
DATABASE_URL=postgresql+psycopg2://biocirv_user:biocirv_dev_password@localhost:5432/biocirv_db
```

Let me know if this helps. Seemed to work for me
